README.md

[![Docker Repository on Quay.io](https://quay.io/repository/tiramisu/mr.docs/status "Docker Repository on Quay.io")](https://quay.io/repository/tiramisu/mr.docs)
[![Documentation Status](https://readthedocs.org/projects/mrdocs/badge/?version=latest)](http://mrdocs.readthedocs.org/en/latest/?badge=latest)

# mr.docs

mr.docs is a [Docker container](https://docker.com "Homepage of docker"), which can help you with testing documentation written in [Sphinx](http://sphinx-doc.org/). 

mr.docs is written as a personal tool, to improve the [documentation of Plone](http://docs.plone.org).

It comes with a pre-configured list 'of good known words' based on typical terms used within [Plone](https://plone.org). 

## Features

- Small Image, based on [Alpine Linux](http://www.alpinelinux.org/).
- You do not need to install sphinx locally.
- Pre-configured with different add-ons.
- A new, clean and 'fresh' build with every test run.
- Various checks


## Documentation

Full documentation for end users can be found in the "docs" folder.

You can find the documentation online on [Read The Docs](https://mrdocs.readthedocs.org/en/latest/).


## Dependencies

- [Docker](https://docker.com "Homepage of docker")

## Installation

Pull the image:

    docker pull quay.io/tiramisu/mr.docs


## Contribute

- [Issue Tracker](github.com/tiramisusolutions/mr.docs/issues)
- [Source Code](github.com/tiramisusolutions/mr.docs)

## Support

If you are having issues, please let us know.


## License

The project is licensed under the GPLv2.


CHANGES.md
# Changelog

## 2.1
- fix wrapper script
- improve Dockerfile
- remove old add-ons
- add coala for testing
- update docs
- update make targets
- add make target to create sphinx config

## 2.0
- update to Alpine 3.4 [svx]
- remove sudo dependency [svx]
- fix spell-check [svx]
- rewrite Dockerfile [svx]
- pin all versions we install with pip [svx]
- rename '.dockerfunc' -> '.mrdocsfunc' [svx]
- rewrite 'mrdocsfunc' [svx]
- add 'help function' to '.mrdocsfunc' [svx]
- change to sphinx rtd theme [svx]

## 0.0.1-dev
- add sphinx-extension: sphinxcontrib-osexample [svx]
- add alabaster as theme [svx]
- update sphinx config for test builds [svx]



tests/doc-checks.sh
#!/bin/bash

# We like color output
# Colors
red=`tput setaf 1`
green=`tput setaf 2`
yellow=`tput setaf 3`
reset=`tput sgr0`


changed_files="$(git diff HEAD~1 --name-only | grep ".*\.rst$")"

#echo "$changed_files" | grep --quiet "$1" && eval "$2"
echo "$changed_files"
if [[ "$changed_files" == *.rst ]]
then
    echo "${green} DUDE ! Awesome we found some .rst files in your last commit, lets run tests against them!${reset}"
    echo "${yellow} Start running spell, link and html checks, please be patient ....!${reset}"
#else
#    echo "rst not found."
fi



src/createc.sh
#!/bin/bash
set -e

# Colors
ESC_SEQ="\x1b["
COL_RESET=$ESC_SEQ"39;49;00m"
COL_RED=$ESC_SEQ"31;01m"
COL_GREEN=$ESC_SEQ"32;01m"
COL_YELLOW=$ESC_SEQ"33;01m"
COL_BLUE=$ESC_SEQ"34;01m"
#COL_MAGENTA=$ESC_SEQ"35;01m"
#COL_CYAN=$ESC_SEQ"36;01m"


## Functions we use
# Error

function error_exit {
	echo -en "$COL_RED$1$COL_RESET" 1>&2
	exit 1
}

#####
ask() {
    # http://djm.me/ask
    while true; do

        if [ "${2:-}" = "Y" ]; then
            prompt="Y/n"
            default=Y
        elif [ "${2:-}" = "N" ]; then
            prompt="y/N"
            default=N
        else
            prompt="y/n"
            default=
        fi

        # Ask the question (not using "read -p" as it uses stderr not stdout)
        echo -en "$COL_BLUE$1$COL_RESET [$prompt] "

        # Read the answer (use /dev/tty in case stdin is redirected from somewhere else)
        read REPLY </dev/tty

        # Default?
        if [ -z "$REPLY" ]; then
            REPLY=$default
        fi

        # Check if the reply is valid
        case "$REPLY" in
            Y*|y*) return 0 ;;
            N*|n*) return 1 ;;
        esac

    done
}
## Example
# Default to No if the user presses enter without giving an answer:
#if ask "Do you want to do such-and-such?" N; then
#    echo "Yes"
#else
#    echo "No"
#fi


####
# Vars
YEAR=$(date +"%Y")

# Use "-d" to skip questions and just use the defaults
while getopts ":d" opt; do
    case "$opt" in
	d)
	    echo -en "$COL_BLUE Using default settings$COL_RESET"
	    exit 0
	;;
    esac
done


# Start running checks, like do we have already d conf.py in docs and so on
echo -en "$COL_YELLOW Running checks ...$COL_RESET"

# If we have no .jekyll we will create one
echo -en "$COL_YELLOW Do we have a .jekyll already? ...$COL_RESET"
if [ -f "docs/.jekyll" ]; then
	echo -en "$COL_YELLOW Found one moving on ...$COL_RESET"
	: # do nothing and move on
else
	echo -en "$COL_YELLOW Creating .jekyll ...$COL_RESET"
	touch docs/.jekyll
fi

# Check if we have already a index.rst if not create one
echo -en "$COL_YELLOW Checking for a index.rst ...$COL_RESET"
if [ -f "docs/index.rst" ]; then
	echo -en "$COL_YELLOW It seems there is no index, creating one now ...$COL_RESET"
	cp src/index.ini docs/index.rst
else
	: # we have one, so we move on
fi

# Check if we have already a config, file, if so tell
if [ -f docs/conf.py ]; then
    	if ask "Do you want to override?" N; then
		rm docs/conf.py
	else
		error_exit "Detected conf.py !  Aborting."
	fi
else
    echo -en "$COL_YELLOW Looking good, lets continue$COL_RESET"
fi


# Start the questions
# Ask for the name of the project
echo -en "$COL_GREEN Name of the project:$COL_RESET"
read project_name
project=$project_name
: "${project:="createc"}"

# Ask for the name of the author
echo -en "$COL_GREEN Name of the author:$COL_RESET"
read project_author
author=$project_author
: "${author:="sven"}"

# Ask for the version, like 1.0
echo -en "$COL_GREEN This is version:$COL_RESET"
read project_version
version=$project_version
: "${version:="1.0"}"

# Ask for the html title
echo -en "$COL_GREEN Please enter a title for html:$COL_RESET"
read project_html_title
project_title=$project_html_title
: "${project_title="My long html title"}"

# Ask for the short title
echo -en "$COL_GREEN Please enter a short title for html:$COL_RESET"
read project_html_short_title
project_title_short=$project_html_short_title
: "${project_title_short="My short title"}"

# Copy the template over
cp src/conf.py.ini docs/conf.py

# Run sed to replace the place holder
sed -i "s/template_project_name/$project/g" docs/conf.py
sed -i "s/template_author/$author/g" docs/conf.py
sed -i "s/template_year/$YEAR/g" docs/conf.py
sed -i "s/template_version/$version/g" docs/conf.py
sed -i "s/template_html.title/$project_title/g" docs/conf.py
sed -i "s/template_html_short_title/$project_title_short/g" docs/conf.py


scripts/copy_config_template.sh
#!/bin/bash
set -e

# Colors
ESC_SEQ="\x1b["
COL_RESET=$ESC_SEQ"39;49;00m"
COL_RED=$ESC_SEQ"31;01m"
COL_GREEN=$ESC_SEQ"32;01m"
COL_YELLOW=$ESC_SEQ"33;01m"
COL_BLUE=$ESC_SEQ"34;01m"
#COL_MAGENTA=$ESC_SEQ"35;01m"
#COL_CYAN=$ESC_SEQ"36;01m"


## Functions we use
# Error

function error_exit {
	echo -en "$COL_RED$1$COL_RESET" 1>&2
	exit 1
}

####
# Vars
YEAR=$(date +"%Y")


# Start running checks, like do we have already d conf.py in docs and so on
echo -en "$COL_YELLOW Running checks ...$COL_RESET\n"

# If we have no .jekyll we will create one
echo -en "$COL_YELLOW Do we have a .jekyll already? ...$COL_RESET\n"
if [ -f "docs/.jekyll" ]; then
	echo -en "$COL_YELLOW Found one moving on ...$COL_RESET\n"
	: # do nothing and move on
else
	echo -en "$COL_YELLOW Creating .jekyll ...$COL_RESET\n"
	touch docs/.jekyll
fi

# Check if we have already a index.rst if not create one
echo -en "$COL_YELLOW Checking for a index.rst ...$COL_RESET\n"
if [ -f "docs/index.rst" ]; then
	echo -en "$COL_YELLOW Found index, moving on ...$COL_RESET\n"
	: # Do nothing
else
	echo -en "$COL_YELLOW Creating example index ...$COL_RESET\n"
	cp templates/index.ini docs/index.rst
fi

# Check if we have already a conf.py, if yes, soft exit
if [ -f docs/conf.py ]; then
	echo -en "$COL_RED You have already a config file !$COL_RESET\n"
	exit 0
else
	echo -en "$COL_YELLOW Copy example configuration ...$COL_RESET\n"
	cp templates/conf.py.ini docs/conf.py
fi

# Check if we already have a _static dir, if not create
echo -en "$COL_YELLOW Checking for _static ...$COL_RESET\n"
if [ -f "docs/_static" ]; then
	echo -en "$COL_YELLOW Found _static moving on ...$COL_RESET\n"
	: # Do nothing
else
	echo -en "$COL_YELLOW Creating _static ...$COL_RESET\n"
	mkdir -p docs/_static
fi

echo
echo -en "$COL_BLUE Done ! \n
 Please adjust settings like ExampleProjectName and ExampleAuthor \n
 in the config under /docs/config.py$COL_RESET \n"

exit 0


scripts/create_config.sh
#!/bin/bash
set -e

# Colors
ESC_SEQ="\x1b["
COL_RESET=$ESC_SEQ"39;49;00m"
COL_RED=$ESC_SEQ"31;01m"
COL_GREEN=$ESC_SEQ"32;01m"
COL_YELLOW=$ESC_SEQ"33;01m"
COL_BLUE=$ESC_SEQ"34;01m"
#COL_MAGENTA=$ESC_SEQ"35;01m"
#COL_CYAN=$ESC_SEQ"36;01m"


## Functions we use
# Error

function error_exit {
	echo -en "$COL_RED$1$COL_RESET" 1>&2
	exit 1
}

#####
ask() {
    # http://djm.me/ask
    while true; do

        if [ "${2:-}" = "Y" ]; then
            prompt="Y/n"
            default=Y
        elif [ "${2:-}" = "N" ]; then
            prompt="y/N"
            default=N
        else
            prompt="y/n"
            default=
        fi

        # Ask the question (not using "read -p" as it uses stderr not stdout)
        echo -en "$COL_BLUE$1$COL_RESET [$prompt] "

        # Read the answer (use /dev/tty in case stdin is redirected from somewhere else)
        read REPLY </dev/tty

        # Default?
        if [ -z "$REPLY" ]; then
            REPLY=$default
        fi

        # Check if the reply is valid
        case "$REPLY" in
            Y*|y*) return 0 ;;
            N*|n*) return 1 ;;
        esac

    done
}
## Example
# Default to No if the user presses enter without giving an answer:
#if ask "Do you want to do such-and-such?" N; then
#    echo "Yes"
#else
#    echo "No"
#fi


####
# Vars
YEAR=$(date +"%Y")

# Use "-d" to skip questions and just use the defaults
while getopts ":d" opt; do
    case "$opt" in
	d)
	    echo -en "$COL_BLUE Using default settings$COL_RESET"
	    exit 0
	;;
    esac
done


# Start running checks, like do we have already d conf.py in docs and so on
echo -en "$COL_YELLOW Running checks ...$COL_RESET"

# If we have no .jekyll we will create one
echo -en "$COL_YELLOW Do we have a .jekyll already? ...$COL_RESET"
if [ -f "docs/.jekyll" ]; then
	echo -en "$COL_YELLOW Found one moving on ...$COL_RESET"
	: # do nothing and move on
else
	echo -en "$COL_YELLOW Creating .jekyll ...$COL_RESET"
	touch docs/.jekyll
fi

# Check if we have already a index.rst if not create one
echo -en "$COL_YELLOW Checking for a index.rst ...$COL_RESET"
if [ -f "docs/index.rst" ]; then
	echo -en "$COL_YELLOW It seems there is no index, creating one now ...$COL_RESET"
	cp templates/index.ini docs/index.rst
else
	: # we have one, so we move on
fi

# Check if we have already a config, file, if so tell
if [ -f docs/conf.py ]; then
    	if ask "Do you want to override?" N; then
		rm docs/conf.py
	else
		error_exit "Detected conf.py !  Aborting."
	fi
else
    echo -en "$COL_YELLOW Looking good, lets continue$COL_RESET"
fi


# Start the questions
# Ask for the name of the project
echo -en "$COL_GREEN Name of the project:$COL_RESET"
read project_name
project=$project_name
: "${project:="ExampleDocs"}"

# Ask for the name of the author
echo -en "$COL_GREEN Name of the author:$COL_RESET"
read project_author
author=$project_author
: "${author:="SomePerson"}"

# Ask for the version, like 1.0
echo -en "$COL_GREEN This is version:$COL_RESET"
read project_version
version=$project_version
: "${version:="1.0"}"

# Ask for the html title
echo -en "$COL_GREEN Please enter a title for html:$COL_RESET"
read project_html_title
project_title=$project_html_title
: "${project_title="My long html title"}"

# Ask for the short title
echo -en "$COL_GREEN Please enter a short title for html:$COL_RESET"
read project_html_short_title
project_title_short=$project_html_short_title
: "${project_title_short="My short title"}"

# Copy the template over
cp templates/conf.py.ini docs/conf.py

# Run sed to replace the place holder
sed -i "s/template_project_name/$project/g" docs/conf.py
sed -i "s/template_author/$author/g" docs/conf.py
sed -i "s/template_year/$YEAR/g" docs/conf.py
sed -i "s/template_version/$version/g" docs/conf.py
sed -i "s/template_html.title/$project_title/g" docs/conf.py
sed -i "s/template_html_short_title/$project_title_short/g" docs/conf.py
